### PR TITLE
Selector disambiguation

### PR DESCRIPTION
**Pull request by silvanshade**
Addresses the discussion starting from https://github.com/madsmtm/objc2/pull/336#issuecomment-1382053237

### CODE REVIEW

File: crates/header-translator/src/cache.rs [hunk 1]
```diff
@@ -119,6 +119,36 @@ impl<'a> Cache<'a> {
...
+                    for this in methods.iter_mut() {
+                        let key = (id.clone(), this.fn_name.clone());
+                        if let Some(that) = names.get_mut(&key) {
+                            that.fn_name = that.selector.replace(':', "_");
+                            this.fn_name = this.selector.replace(':', "_");
+                        } else {
+                            names.insert(key, this);
+                        }
+                    }
+                }
```
Comment (madsmtm) [inline crates/header-translator/src/cache.rs:147]: You can avoid the duplication by matching like this:
```rust
Stmt::Methods { cls: id, methods, .. }
| Stmt::ProtocolDecl { id, methods, .. } =>
```

File: crates/header-translator/src/cache.rs [hunk 2]
```diff
@@ -119,6 +119,36 @@ impl<'a> Cache<'a> {
     }
 
     fn update_file(&self, file: &mut File) {
+        // disambiguate duplicate names
```
Comment (madsmtm) [inline crates/header-translator/src/cache.rs:122]: Nit: Add a comment that this only works within a single file (which is fine to start with!)

File: crates/header-translator/src/cache.rs [hunk 3]
```diff
@@ -119,6 +119,36 @@ impl<'a> Cache<'a> {
...
+                Stmt::Methods { cls, methods, .. } => {
+                    for this in methods.iter_mut() {
+                        let key = (cls.clone(), this.fn_name.clone());
+                        if let Some(that) = names.get_mut(&key) {
+                            that.fn_name = that.selector.replace(':', "_");
+                            this.fn_name = this.selector.replace(':', "_");
+                        } else {
+                            names.insert(key, this);
+                        }
+                    }
```
Comment (madsmtm) [inline crates/header-translator/src/cache.rs:135]: Nit: A bit better naming IMO?

```suggestion
                    for method in methods.iter_mut() {
                        let key = (cls.clone(), method.fn_name.clone());
                        if let Some(other) = names.get_mut(&key) {
                            other.fn_name = other.selector.replace(':', "_");
                            method.fn_name = method.selector.replace(':', "_");
                        } else {
                            names.insert(key, method);
                        }
                    }
```

**Review by madsmtm — COMMENTED** [review]
Cool!

**Comment by madsmtm** [comment]
I think I would prefer the `icrate-generated` changes to happen in separate commits, to make it easier to see what has changed.

File: crates/header-translator/src/cache.rs (hunk 1 above)
Comment (silvanshade) [reply crates/header-translator/src/cache.rs:147]: Ah, good point!

File: crates/header-translator/src/cache.rs (hunk 2 above)
Comment (silvanshade) [reply crates/header-translator/src/cache.rs:122]: Are there instances where methods for a class or protocol can potentially be split across files? Or are there some other scenarios where we would need to disambiguate?

**Comment by silvanshade** [comment]
> Also, I think I would prefer the `icrate-generated` changes to happen in separate commits, to make it easier to see what has changed.

Sure, I can do that. The reason I included it here is because I was expecting the CI to fail otherwise with a complaint about changes in the generated output.

File: crates/header-translator/src/cache.rs (hunk 2 above)
Comment (madsmtm) [reply crates/header-translator/src/cache.rs:122]: Indeed there are: Any object can have new "categories" defined on it at any time. But it likely won't happen, so again, that it only works within a single file is fine.

**Comment by madsmtm** [comment]
I think I meant that I would like the commit that you made in `icrate-generated` to be split into two, not that the commit should be decoupled from this PR - sorry for the confusion!

**Comment by madsmtm** [comment]
I went and rebased your branch to fix it, as well as adding support for instance vs. class methods - I hope you don't mind!

**Pull request merged**